x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/FieldAttribute.java

```diff
@@ -105,6 +107,19 @@ public FieldAttribute(StreamInput in) throws IOException {
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
+        if (out.getTransportVersion().onOrAfter(TransportVersions.ESQL_FIELD_ATTRIBUTE_CACHED_SERIALIZATION)
+            && out instanceof PlanStreamOutput pso) {
+            Integer inCache = pso.fromCache(this);
+            if (inCache != null) {
+                out.writeBoolean(true);
+                out.writeInt(inCache);
+                return;
+            }
+
+            Integer newCacheId = pso.addToCache(this);
```

Comment:
If we add a max cache size, we'll need three-value logic here (new, cached, no-cache). I could encode the three states together with the ID, but it would be pretty confusing, I think I'll use a single byte for the state and a zlong for the id.

Reply:
If we have too many attributes to put in the cache, don't we also have too many attributes to send over the wire?

Meaning: once we reach the cache's limit, shouldn't we rather fail and tell the user "woah, whatever you're trying to do, it's too big"?

Reply:
Yeah, I was thinking something similar.
IMHO we should not fail because of cache size though, I'd really prefer to circuit break based on the size of the serialization buffer (for sure on the coordinator, maybe also on the data node?)

Still, the concern on the max cache size remains: my feeling is that limiting the cache would just make things worse in the worst scenarios.

Reply:
I think I completely misunderstood the proposal about limiting the cache size. 
I'll just remove the no-cache option and I'll throw an error if we exceed the cache size